witchcraft/server_metrics.go

```diff
@@ -118,6 +147,106 @@ func (s *Server) initMetrics(ctx context.Context, installCfg config.Install) (rR
+// otherwise.
+func zeroValuesForMetricType(metricType string) map[string]interface{} {
+	var zeroMetric interface{}
+	switch metricType {
+	case "meter":
+		zeroMetric = gometrics.NewMeter()
+	case "timer":
+		zeroMetric = gometrics.NewTimer()
+	case "histogram":
+		zeroMetric = gometrics.NewHistogram(metrics.DefaultSample())
+	default:
+		panic(fmt.Sprintf("should never happen: unsupported metricType %s", metricType))
+	}
+	return metrics.ToMetricVal(zeroMetric).Values()
+}
+
+// logZeroValueMetric logs a "zero" value for metric with the provided metricID, metricType, and tags to the provided
+// metricLogger if an entry for the metric has not yet been logged (which is determined based on whether or not an entry
+// for the metric exists in the provided seenMetrics set). Returns the zero values of the logged metric. If a zero value
+// is logged, an entry for the metric is added to seenMetrics.
+func logZeroValueMetric(
+	metricLogger metric1log.Logger,
+	metricID,
+	metricType string,
+	tags map[string]string,
+	seenMetrics *seenMetricsSet,
+	metricTypeValuesDisallowedList map[string]map[string]struct{},
+	metricTagsParam metric1log.Param,
+) (zeroValuesLogged map[string]interface{}) {
+
+	// if metric has been seen before, no need to log zero value
+	mapKey := tagMapKey(metricID, metricType, tags)
+
+	// only acquire read lock while checking because this should be true vast majority of the time
+	seenMetrics.lock.RLock()
+	_, metricSeen := seenMetrics.seenSet[mapKey]
+	seenMetrics.lock.RUnlock()
+	if metricSeen {
+		return nil
+	}
+
+	zeroValuesToUse := zeroValuesForMetricType(metricType)
+	removeDisallowedKeys(metricType, zeroValuesToUse, metricTypeValuesDisallowedList)
+
+	// if removing the disallowed keys causes the values to be empty, no need to log zero value
+	if len(zeroValuesToUse) == 0 {
+		return nil
+	}
+
+	seenMetrics.lock.Lock()
+	// need to check again while under lock because read lock was only held during initial check to optimize for fact
+	// that it will return true vast majority of the time. However, want to ensure that only one zero-value log line is
+	// emitted, so check condition again while holding write lock.
+	if _, ok := seenMetrics.seenSet[mapKey]; !ok {
+		// metric not seen before: emit zero-value and record
+		metricLogger.Metric(metricID, metricType, metric1log.Values(zeroValuesToUse), metricTagsParam)
+		seenMetrics.seenSet[mapKey] = struct{}{}
+		zeroValuesLogged = zeroValuesToUse
+	}
+	seenMetrics.lock.Unlock()
+	return zeroValuesLogged
+}
+
+// tagMapKey returns a string that provides a unique key for the provided metricID, metricType and tagsMap. The content
+// of the map is put into a slice where each element is a string of the form "{key}:{value}" and the slice is then
+// sorted. The tags map (rather than slice) is used to construct the key because the map is what the metric logger
+// ultimately emits.
+func tagMapKey(metricID, metricType string, tagsMap map[string]string) string {
```

Comment:
Yes, understood. However, not totally sure how to improve on this unless we assume that tag order will be stable for a given metric/we will treat different tag ordering differently (I don't think we can just print the map directly since map iteration order is random).

One thing to note is that the penalty/overhead for metric logging should be lower than for service logging -- this call is basically made for the number of matching metric types each "emit frequency" unit of time (unlike service logs, which are emitted on every call to the logger)

Reply:
Yeah, I am not sure how to optimize this without further changes and I believe it will be fine, but just wanted to flag in case we do see any problems. Good point on the `emit-frequency` which certainly amortizes the cost.